src/main/java/com/shapesecurity/salvation/Parser.java

```diff
@@ -151,208 +206,335 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
+        if (this.eat(DirectiveValueToken.class)) {
+            this.error("The " + tokenValue + " directive must not contain any value");
+            throw NON_EMPTY_VALUE_TOKEN_LIST;
+        }
+    }
+
+    @Nonnull private Set<MediaType> parseMediaTypeList() throws DirectiveParseException {
         Set<MediaType> mediaTypes = new LinkedHashSet<>();
-        if (!this.hasNext(DirectiveValueToken.class)) {
-            throw this.createError("media-type-list must contain at least one media-type");
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            for (SubDirectiveValueToken subdv : splitByWSP(dv)) {
+                try {
+                    mediaTypes.add(this.parseMediaType(subdv));
+                } catch (DirectiveValueParseException e) {
+                    parseException = true;
+                }
+            }
+            if (parseException) {
+                throw INVALID_MEDIA_TYPE_LIST;
+            }
         }
-        mediaTypes.add(this.parseMediaType());
-        while (this.hasNext(DirectiveValueToken.class)) {
-            mediaTypes.add(this.parseMediaType());
+        if (mediaTypes.isEmpty()) {
+            this.error("The media-type-list must contain at least one media-type");
+            throw INVALID_MEDIA_TYPE_LIST;
         }
         return mediaTypes;
     }
 
-    @Nonnull private MediaType parseMediaType() throws ParseException {
-        Token token = this.advance();
+    @Nonnull private MediaType parseMediaType(@Nonnull SubDirectiveValueToken token) throws DirectiveValueParseException {
         Matcher matcher = Constants.mediaTypePattern.matcher(token.value);
         if (matcher.find()) {
             return new MediaType(matcher.group("type"), matcher.group("subtype"));
         }
-        throw this.createError("expecting media-type but found " + token.value);
+        this.error(token, "Expecting media-type but found " + token.value);
+        throw INVALID_MEDIA_TYPE;
     }
 
-    @Nonnull private Set<SourceExpression> parseSourceList() throws ParseException {
+    @Nonnull private Set<SourceExpression> parseSourceList() throws DirectiveParseException {
         Set<SourceExpression> sourceExpressions = new LinkedHashSet<>();
-        if (this.hasNext(DirectiveValueToken.class) && this.peek().value
-            .equalsIgnoreCase("'none'")) {
-            this.advance();
-            sourceExpressions.add(None.INSTANCE);
-            return sourceExpressions;
-        }
-        while (this.hasNext(DirectiveValueToken.class)) {
-            sourceExpressions.add(this.parseSourceExpression());
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            if (trimRHSWS(dv.value).equalsIgnoreCase("'none'")) {
```

Comment:
We can get rid of this exceptional condition by keeping track of the `None` values returned by `parseSourceExpression` in the loop below. I think that would be more DRY.